components/locid_transform/src/directionality.rs

```diff
@@ -0,0 +1,153 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::provider::*;
+use crate::{LocaleExpander, LocaleTransformError};
+use icu_locid::{subtags_script as script, Locale};
+use icu_provider::{DataPayload, DataProvider};
+
+/// Represents the direction of a script.
+///
+/// [`LocaleDirectionality`] can be used to get this information.
+#[zerovec::make_ule(DirectionULE)]
+#[zerovec::derive(Debug)]
+#[repr(u8)]
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+#[cfg_attr(
+    feature = "datagen",
+    derive(serde::Serialize, databake::Bake),
+    databake(path = icu_locid_transform::provider),
+)]
+#[cfg_attr(feature = "serde", derive(serde::Deserialize))]
+#[non_exhaustive]
+pub enum Direction {
+    /// The script is left-to-right.
+    LeftToRight = 0,
+    /// The script is right-to-left.
+    RightToLeft = 1,
+    /// The direction of the script is unknown or undefined.
+    Unknown = 2,
+}
+
+/// The `LocaleDirectionality` provides methods to determine the direction of a locale based
+/// on [`CLDR`] data.
+///
+/// # Examples
+///
+/// ```
+/// use icu_locid::Locale;
+/// use icu_locid_transform::LocaleDirectionality;
+///
+/// let ld = LocaleDirectionality::try_new_unstable(&icu_testdata::unstable())
+///     .expect("create failed");
+///
+/// let locale: Locale = "en".parse().unwrap();
+/// assert_eq!(ld.get(locale), icu_locid_transform::Direction::LeftToRight);
+/// ```
+///
+/// [`CLDR`]: http://cldr.unicode.org/
+#[derive(Debug)]
+pub struct LocaleDirectionality {
+    rtl: DataPayload<DirectionalityV1Marker>,
+    expander: LocaleExpander,
+}
+
+impl LocaleDirectionality {
+    /// A constructor which takes a [`DataProvider`] and creates a [`LocaleDirectionality`].
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    /// <div class="stab unstable">
+    /// ⚠️ The bounds on this function may change over time, including in SemVer minor releases.
+    /// </div>
+    pub fn try_new_unstable<P>(provider: &P) -> Result<LocaleDirectionality, LocaleTransformError>
+    where
+        P: DataProvider<DirectionalityV1Marker>
+            + DataProvider<LikelySubtagsForLanguageV1Marker>
+            + DataProvider<LikelySubtagsForScriptRegionV1Marker>
+            + ?Sized,
+    {
+        let expander = LocaleExpander::try_new_unstable(provider)?;
+        Self::try_new_with_expander_unstable(provider, expander)
+    }
+
+    icu_provider::gen_any_buffer_constructors!(
+        locale: skip,
+        options: skip,
+        error: LocaleTransformError
+    );
+
+    // TODO: Do we need to allow user-provided LocaleExpanders? Can a LocaleExpander ever not
+    // have the right amount of data? Make this public if so
+    fn try_new_with_expander_unstable<P>(
```

Comment:
Question: I _think_ we don't need to make this public, because both `DirectionalityV1Marker` and the `LikelySubtagsForX` only support loading the full set of data (by specifying the `und` locale). 

The same applies to `LocaleCanonicalizer`, though, so I don't understand what this comment means: https://github.com/unicode-org/icu4x/blob/main/components/locid_transform/src/canonicalizer.rs#L251

Reply:
`LocaleExpander::try_new_extended_unstable` loads `LikelySubtagsExtendedV1Marker` in addition to the core data, and the user can decide which one to use because the extended data is quite big (long tail languages). You should support both variants by making this function public.

Reply:
I see - would this ever make a difference for the directionality, or do we just want to have this to allow already-constructed extended LocaleExpanders to be reusable here?

Reply:
If you're asking whether you should split the directionality key into two - no I don't think that's necessary, as the numbers of scripts is quite a lot smaller than the number of languages (https://github.com/unicode-org/icu4x/blob/main/provider/repodata/data/json/locid_transform/likelysubtags_ext%401/und.json is huge).

You do however need this data for maximizing, which is why you should let the client choose the expander. You could alternatively have a core and an extended constructor, but that's not how `LocaleCanonicalizer` does it, so we should follow that lead.

Reply:
My question is what client code would break/return different results if we did not allow the extended LocaleExpander to be used, and only ever used the LocaleExpander with core data (i.e., the way it is now with the private expander constructor)?

Reply:
It would not be able to determine the script for e.g. the `jbn` language, so it would return `Unknown` instead of `RightToLeft`.

Reply:
Thanks!